File: 2022/d02.zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    const file = try std.fs.cwd().openFile("d02.in", .{ .mode = .read_only });
    defer file.close();

    const contents = try file.reader().readAllAlloc(allocator, 10 * 1024 * 1024);
    defer allocator.free(contents);

    var split_iter = std.mem.split(u8, contents, "\n");
    var points: i32 = 0;
    var points2: i32 = 0;
    while (split_iter.next()) |line| {
        if (line.len == 0) continue;

        const opponent = @intCast(i32, line[0] - 'A');
        const me = @intCast(i32, line[2] - 'X');
        // see cpp for explanation
        const outcome = @mod(me - opponent + 1, 3);

        points += outcome * 3;
        points += me + 1;

        // part2
        const needed_outcome = me;
        const needed_me: i32 = @mod(opponent - 1 + needed_outcome, 3);
        points2 += needed_outcome * 3 + needed_me + 1;
    }

    std.debug.print("Part1: Total points are {}\n", .{points});
    std.debug.print("Part2: Total points are {}\n", .{points2});
}

File: 2022/rust/src/bin/d02.rs
use std::fs;
use std::io;

pub fn main() -> Result<(), io::Error> {
    let contents = fs::read_to_string("../d02.in")?;

    let mut points = 0;
    let mut points2 = 0;
    for line in contents.split('\n') {
        if line == "" { continue };
        // println!("{}", line);

        let line_bytes = line.as_bytes();
        let opponent = (line_bytes[0] - b'A') as i32;
        let me = (line_bytes[2] - b'X') as i32;
        // see cpp for explanation
        // rem_euclid won't have negative outcomes when doing 'modulo'
        let outcome = (me - opponent + 1).rem_euclid(3);
        points += outcome * 3;
        points += me + 1;

        // part2
        let needed_outcome = me;
        let needed_me = (opponent - 1 + needed_outcome).rem_euclid(3);
        points2 += needed_outcome * 3 + needed_me + 1;
    }

    println!("Part1: Total points are {}", points);
    println!("Part2: Total points are {}", points2);

    Ok(())
}
